Support running Ubuntu image on GCE node

yguo0905 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#pull-request-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

This PR (on top of #44629) contains the script changes for running Ubuntu image on GCE.

**Special notes for your reviewer**:

We made change in `gci/node.yaml` and `gci/master.yaml` to ensure that Kubernetes jobs can start automatically after reboot. This is not needed for GCI but required by Ubuntu. See https://github.com/kubernetes/kubernetes/pull/44744#discussion_r113105970 for details. With this change, Ubuntu could use the same provisioning scripts as GCI's.

Ran e2e tests using the following command and all tests passed.

```
KUBE_GCE_NODE_IMAGE=ubuntu-gke-1604-xenial-v20170420-1 KUBE_GCE_NODE_PROJECT=ubuntu-os-gke-cloud KUBE_NODE_OS_DISTRIBUTION=ubuntu GINKGO_PARALLEL=y GINKGO_PARALLEL_NODES=30 go run hack/e2e.go -- -v --build --up --test --test_args="--ginkgo.skip=\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]" --down
```

Also tested manually for both GCI and Ubuntu images.

**Running Ubuntu image on master is not in the scope and has not been tested.**


**Release note**:
`Support Ubuntu 16.04 image on GCE`

Commit:
Add Ubuntu as a platform for e2e tests on GCE

Using Ubuntu on GCE to run cluster e2e tests requires slightly different
node.yaml and master.yaml files than GCI, because Ubuntu uses systemd as
PID 1, wheras GCI uses upstart with a systemd delegate. Therefore the
e2e tests fail using those files since the kubernetes services are not
brought back up after a node/master reboot.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44744)
<!-- Reviewable:end -->

yguo0905 commented:
@chrisglass 
@dchen1107

chrisglass commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -1,128 +0,0 @@
-      Description=Kubernetes
-
-      [Install]
-      WantedBy=multi-user.target
```
This is needed by Ubuntu.

The GCI image doesn't have this in its systemd services because GCI uses systemd as a delegate (pid 1 is upstart).

Therefore symlinking ubuntu -> gci does not work: if a node reboots the services won't be brought back up.

That is the reason I originally duplicated some of the files in the subdirectory instead of symlinking, the node.yaml and master.yaml files need to be different.

Or does this imply that GCI will change to use systemd as pid 1?

chrisglass commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -1,128 +0,0 @@
- - systemctl enable kubelet-monitor.service
- - systemctl enable kube-logrotate.timer
- - systemctl enable kube-logrotate.service
- - systemctl enable kubernetes.target
```
This is needed by Ubuntu.

This extra "enable" call ensures that the appropriate symlink is created by systemd (so that it is brought back up after a reboot).

chrisglass commented on cluster/gce/ubuntu/node-helper.sh:
```diff
@@ -1,32 +0,0 @@
-  ensure-gci-metadata-files
-  create-node-template "$template_name" "${scope_flags[*]}" \
-    "kube-env=${KUBE_TEMP}/node-kube-env.yaml" \
-    "user-data=${KUBE_ROOT}/cluster/gce/ubuntu/node.yaml" \
```
The existence of this file was to make sure the approrpiate node.yaml file is passed to the node instance (an ubuntu specific one, not the GCI one).

chrisglass commented on cluster/gce/ubuntu/master.yaml:
```diff
@@ -1,128 +0,0 @@
- - systemctl enable kubelet-monitor.service
- - systemctl enable kube-logrotate.timer
- - systemctl enable kube-logrotate.service
- - systemctl enable kubernetes.target
```
This is needed by Ubuntu.

This extra "enable" call ensures that the appropriate symlink is created by systemd (so that it is brought back up after a reboot).

chrisglass commented on cluster/gce/ubuntu/master.yaml:
```diff
@@ -1,128 +0,0 @@
-      Description=Kubernetes
-
-      [Install]
-      WantedBy=multi-user.target
```
This is needed by Ubuntu.

The GCI image doesn't have this in its systemd services because GCI uses systemd as a delegate (pid 1 is upstart).

Therefore symlinking ubuntu -> gci does not work: if a node reboots the services won't be brought back up.

That is the reason I originally duplicated some of the files in the subdirectory instead of symlinking, the node.yaml and master.yaml files need to be different.

chrisglass reviewed (commented):
I'm unsure why the ubuntu-specific subdirectory was deleted and instead synlinked to the GCI one, since in my experience this doesn't work.

Specifically, the kubernetes services are not restarted after a system reboot since the systemd target is not enabled, nor does it have an [Install]\nWantedBy=<something> entry.

Or is the assumption that the result of some of the tests should be ignored? (at least one of them uses rebooting nodes to simulate network partition) 
Is it assumed that user will never reboot nodes?

yguo0905 commented:
You are right. I wasn't aware of the reboot issue. The tests I ran (the slow, serial, disruptive, flaky and feature tests were skipped) didn't cover this case.
I will keep the original commit and set up Jenkins jobs to run more tests against it.

spxtr commented:
@k8s-bot test this

chrisglass commented:
Since this is basically the same as #44629 and I don't have push access to this branch, I suggest this to be closed and efforts focused on #44629 instead.

dchen1107 commented on cluster/gce/ubuntu/master-helper.sh:
```diff
@@ -0,0 +1,111 @@
+    --tags "${MASTER_TAG}" \
+    --network "${NETWORK}" \
+    --scopes "storage-ro,compute-rw,monitoring,logging-write" \
+    --can-ip-forward \
```
@chrisglass By comparing gce/ubuntu/master-helper.sh and gce/gci/master-helper.sh, adding --can-ip-forward upon creating the instance is the only real difference here. Why do we need this if GCI / COS doesn't require it? cc/ @roberthbailey @zmerlynn 

The reason I care this subtle difference is that I want to make sure the difference between two package as minimal as possible.

chrisglass replied:
I suspect I "forked" the script when GCI still had --can-ip-forward. I am fine with dropping it.

roberthbailey replied:
I'm surprised that COS doesn't require the --can-ip-forward flag.

@bowei

bowei replied:
@dchen1107 -- Is this because we never contact a pod IP on the master from any other node?

yguo0905 replied:
Both [master-helper.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/master-helper.sh#L90) and [node-helper.sh (via `create-node-template`)](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/util.sh#L524) call `make-gcloud-network-argument`, which [sets](https://github.com/kubernetes/kubernetes/blob/eda8e12723b4a544705dae5e86f5d687bb7f22ba/cluster/gce/util.sh#L472) --can-ip-forward unless `KUBE_GCE_ENABLE_IP_ALIASES` is set. So we are good.

dchen1107 commented on cluster/gce/ubuntu/master-helper.sh:
```diff
@@ -0,0 +1,111 @@
+
+  echo "${kube_env}" > ${KUBE_TEMP}/master-kube-env.yaml
+  get-metadata "${existing_master_zone}" "${existing_master_name}" cluster-name > "${KUBE_TEMP}/cluster-name.txt"
+  get-metadata "${existing_master_zone}" "${existing_master_name}" gci-update-strategy > "${KUBE_TEMP}/gci-update.txt"
```
GCI specific variable, no need here.

dchen1107 commented on cluster/gce/ubuntu/master-helper.sh:
```diff
@@ -0,0 +1,111 @@
+  echo "${kube_env}" > ${KUBE_TEMP}/master-kube-env.yaml
+  get-metadata "${existing_master_zone}" "${existing_master_name}" cluster-name > "${KUBE_TEMP}/cluster-name.txt"
+  get-metadata "${existing_master_zone}" "${existing_master_name}" gci-update-strategy > "${KUBE_TEMP}/gci-update.txt"
+  get-metadata "${existing_master_zone}" "${existing_master_name}" gci-ensure-gke-docker > "${KUBE_TEMP}/gci-ensure-gke-docker.txt"
```
GCI specific legacy variable, no need here, and should be removed from gci too.

dchen1107 commented on cluster/gce/ubuntu/master-helper.sh:
```diff
@@ -0,0 +1,111 @@
+  get-metadata "${existing_master_zone}" "${existing_master_name}" cluster-name > "${KUBE_TEMP}/cluster-name.txt"
+  get-metadata "${existing_master_zone}" "${existing_master_name}" gci-update-strategy > "${KUBE_TEMP}/gci-update.txt"
+  get-metadata "${existing_master_zone}" "${existing_master_name}" gci-ensure-gke-docker > "${KUBE_TEMP}/gci-ensure-gke-docker.txt"
+  get-metadata "${existing_master_zone}" "${existing_master_name}" gci-docker-version > "${KUBE_TEMP}/gci-docker-version.txt"
```
ditto...

chrisglass commented on cluster/gce/ubuntu/master-helper.sh:
```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+
+# Copyright 2016 The Kubernetes Authors.
```
Changed in https://github.com/kubernetes/kubernetes/pull/44629/commits/f07669ba6c99c02b519addf98459c51f5343d18a

dchen1107 commented on cluster/gce/ubuntu/master.yaml:
```diff
@@ -0,0 +1,128 @@
+ - systemctl enable kubelet-monitor.service
+ - systemctl enable kube-logrotate.timer
+ - systemctl enable kube-logrotate.service
+ - systemctl enable kubernetes.target
```
Based on my understanding, this is the only line (another one in node.yaml) must required to creating a separate gce/ubuntu folder besides gce/common.sh. I am wondering if we can introduce the same:

- systemctl enable kubernetes.target 
before start kubernetes.target. If we could, we can hugely simplify the codebase, and avoid future unnecessary devergening and minimize the maintenance cost.

cc/ @kubernetes/sig-cluster-lifecycle-api-reviews

yguo0905 replied:
I've looked into this. In GCI, the cloud-init's specs in `/var/lib/cloud` are not persisted across reboots, so cloud-init will always run everything as if it's on a newly created machine<sup>1</sup>. Therefore, `systemctl enable kubernetes.target` is useless in this case.

However, this is not the case for Ubuntu, where cloud-init will skip those steps whose frequency is `once`. That's why we need to have `kubernetes.target` required by state `multi-user.target` and have it enabled. But I'm not sure if it needs to be strictly started after `docker.service` though. (BTW, the `After` should be under section `[Unit]` rather than `[Install]`).

Therefore, I think it should be safe to make the changes for GCI and change `ubuntu` to be a symlink to `gci`. Both GCI and Ubuntu should be working with this change. (I'm not sure if Ubuntu should have the same behavior as GCI wrt cloud-init though.)

@roberthbailey and @kubernetes/sig-cluster-lifecycle-api-reviews, can you please confirm?

<br>
<sup>1</sup> This is because the upstart jobs are in /etc/init, which is not persisted across reboots in GCI. In this case, the cloud-init will be rerun in each reboot but it will not recreate the jobs because their run frequency is once. In order to fix this, GCI needs to reset /var/lib/cloud during restart to make cloud-init always run everything.

dchen1107 commented on cluster/gce/ubuntu/node-helper.sh:
```diff
@@ -0,0 +1,32 @@
+    "user-data=${KUBE_ROOT}/cluster/gce/ubuntu/node.yaml" \
+    "configure-sh=${KUBE_ROOT}/cluster/gce/gci/configure.sh" \
+    "cluster-name=${KUBE_TEMP}/cluster-name.txt" \
+    "gci-update-strategy=${KUBE_TEMP}/gci-update.txt" \
```
no need this.

dchen1107 commented on cluster/gce/ubuntu/node-helper.sh:
```diff
@@ -0,0 +1,32 @@
+    "configure-sh=${KUBE_ROOT}/cluster/gce/gci/configure.sh" \
+    "cluster-name=${KUBE_TEMP}/cluster-name.txt" \
+    "gci-update-strategy=${KUBE_TEMP}/gci-update.txt" \
+    "gci-ensure-gke-docker=${KUBE_TEMP}/gci-ensure-gke-docker.txt" \
```
ditto...

dchen1107 commented on cluster/gce/ubuntu/node-helper.sh:
```diff
@@ -0,0 +1,32 @@
+    "cluster-name=${KUBE_TEMP}/cluster-name.txt" \
+    "gci-update-strategy=${KUBE_TEMP}/gci-update.txt" \
+    "gci-ensure-gke-docker=${KUBE_TEMP}/gci-ensure-gke-docker.txt" \
+    "gci-docker-version=${KUBE_TEMP}/gci-docker-version.txt"
```
ditto...

dchen1107 commented on cluster/gce/ubuntu/node.yaml:
```diff
@@ -0,0 +1,128 @@
+ - systemctl enable kubelet-monitor.service
+ - systemctl enable kube-logrotate.timer
+ - systemctl enable kube-logrotate.service
+ - systemctl enable kubernetes.target
```
Same comment as the one in master.yaml

yguo0905 committed:
Makes cluster/gce/ubuntu to be a symlink to cluster/gce/gci and changes the gci's [master|node].yaml to enable kubernetes.target.

This enables Ubuntu to use the same provisioning scripts as GCI's. The
change for enabling kubernetes.target is needed for Ubuntu but a no-op
for GCI.

yguo0905 commented:
I've changed `ubuntu` to be a symlink to `gci` and made changes to start `kubernetes.target` automatically after reboot. Tested manually for both GCI and Ubuntu images (for reboots). Please take another look.

yguo0905 commented:
Besides my manual tests, also confirmed with @adityakali, the changes in [master|node].yaml should be a no-op for GCI.

dchen1107 commented:
I updated the release-notes a little bit: Support Ubuntu 16.04 image on GCE

/lgtm

mikedanese commented on cluster/gce/util.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright 2014 The Kubernetes Authors.
+# Copyright 2017 The Kubernetes Authors.
```
We don't update copyright dates. The dates should be when the file was added.
